README.md
# test_pl

A new Flutter plugin develop for learning purpose

## Getting Started

This project is a starting point for a Flutter
[plug-in package](https://flutter.dev/developing-packages/),
a specialized package that includes platform-specific implementation code for
Android and/or iOS.

For help getting started with Flutter, view our
[online documentation](https://flutter.dev/docs), which offers tutorials,
samples, guidance on mobile development, and a full API reference.



CHANGELOG.md
## 1.0.0

* TODO: Describe initial release.
## Pacakge develop for learning purpose


lib/test_pl.dart
library helpers;

export 'package:test_pl/currency_helper.dart';
export 'package:test_pl/platform_version_helper.dart';


lib/test_pl_web.dart
import 'dart:async';
// In order to *not* need this ignore, consider extracting the "web" version
// of your plugin as a separate package, instead of inlining it in the same
// package as the core of your plugin.
// ignore: avoid_web_libraries_in_flutter
import 'dart:html' as html show window;

import 'package:flutter/services.dart';
import 'package:flutter_web_plugins/flutter_web_plugins.dart';

/// A web implementation of the TestPl plugin.
class TestPlWeb {
  static void registerWith(Registrar registrar) {
    final MethodChannel channel = MethodChannel(
      'test_pl',
      const StandardMethodCodec(),
      registrar,
    );

    final pluginInstance = TestPlWeb();
    channel.setMethodCallHandler(pluginInstance.handleMethodCall);
  }

  /// Handles method calls over the MethodChannel of this plugin.
  /// Note: Check the "federated" architecture for a new way of doing this:
  /// https://flutter.dev/go/federated-plugins
  Future<dynamic> handleMethodCall(MethodCall call) async {
    switch (call.method) {
      case 'getPlatformVersion':
        return getPlatformVersion();
      default:
        throw PlatformException(
          code: 'Unimplemented',
          details: 'test_pl for web doesn\'t implement \'${call.method}\'',
        );
    }
  }

  /// Returns a [String] containing the version of the platform.
  Future<String> getPlatformVersion() {
    final version = html.window.navigator.userAgent;
    return Future.value(version);
  }
}


test/test_pl_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:test_pl/test_pl.dart';

void main() {
  const MethodChannel channel = MethodChannel('test_pl');

  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return '42';
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });

  test('getPlatformVersion', () async {
    expect(await TestPl.platformVersion, '42');
  });
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'dart:async';

import 'package:flutter/services.dart';
import 'package:test_pl/test_pl.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  String _platformVersion = 'Unknown';

  @override
  void initState() {
    super.initState();
    initPlatformState();
  }

  // Platform messages are asynchronous, so we initialize in an async method.
  Future<void> initPlatformState() async {
    // DataHelper().isNullOrBlank('val');
    dynamic checkValue = DataHelper.isNullOrBlank('val');
    print('checkValue, $checkValue');

    String platformVersion;
    // Platform messages may fail, so we use a try/catch PlatformException.
    // We also handle the message potentially returning null.
    try {
      platformVersion = await DataHelper.platformVersion ?? 'Unknown platform version';
    } on PlatformException {
      platformVersion = 'Failed to get platform version.';
    }

    // If the widget was removed from the tree while the asynchronous platform
    // message was in flight, we want to discard the reply rather than calling
    // setState to update our non-existent appearance.
    if (!mounted) return;

    setState(() {
      _platformVersion = platformVersion;
    });
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: Column(
          children: [
            Center(
              child: Text('Running on: $_platformVersion\n'),
            ),
            Center(
              child: Text('Check Value is Null: ${DataHelper.isNullOrBlank('')}\n'),
            ),
            Center(
              child: Text('isInteger("1"): ${DataHelper.isInteger('1')}\n'),
            ),
            Center(
              child: Text('isNumeric("1.2"): ${DataHelper.isNumeric('1')}\n'),
            ),
            Center(
              child: Text('Running on: ${DataHelper.toNumber('100')}\n'),
            ),
            //[10, '', '', null, 20, 0, 40, 50, false]
            Center(
              child: Text('Concat : ${DataHelper.concat([1], [3])}\n'),
            ),
            Center(
              child: Text('Difference From First : ${DataHelper.differenceFromFirst([
                    {'a': 1},
                    2,
                    3,
                    4,
                    5,
                    6,
                    7
                  ], [
                    3,
                    5,
                    4,
                    11,
                    6,
                    7,
                    9
                  ])}\n'),
            ),
            Center(
              child: Text('Difference : ${DataHelper.difference([
                    {'a': 1},
                    2,
                    3,
                  ], [
                    3,
                    5,
                    {'b': 4},
                    11
                  ])}\n'),
            ),
            Center(
              child: Text('remove At : ${DataHelper.removeAt([1, 2, 3, 4, 5], 2)}\n'),
            ),

            Center(
              child: Text('Find Index : ${DataHelper.findIndex([
                    {'abc': 'Nitin2', 't': '1'},
                    {'abc': 'Nitin3', 't': '1'},
                    {'abc': 'Nitin', 't': '1'},
                    {'abc': 'Nitin', 't': '1'},
                    // {'abc': 'Nitin'},
                    // {'abc': 'krushant'},
                  ], {
                    'abc': 'Nitin',
                    't': '1'
                  })}\n'),
            ),

            /// findLastIndex
            Center(
              child: Text('Last Index : ${DataHelper.findLastIndex([
                    {'abc': 'Nitin2'},
                    {'abc': 'Nitin', 'text': '2'},
                    {'abc': 'Nitin', 'text': '1'},
                    {'abc': 'krushant', 'text': '2'},
                    {'abc': 'Nitin', 'text': '2'},
                  ], {
                    'abc': 'Nitin',
                    'text': '2'
                  })}\n'),
            ),

            /// contains
            Center(
              child: Text('Contains : ${DataHelper.contains([1, 2, 3, 4, 6, 7], 5)}\n'),
            ),

            /// containsfromList
            Center(
              child: Text('Contains From List : ${DataHelper.containsfromList({1: 'Jupiter', 2: 'Saturn'}, 'Jupiter')}\n'),
            ),

            /// fold
            ///  ${DataHelper.fold([10, 2, 5, 0.5])}
            ///   ${DataHelper.fold([10, 2, 5, 0.5],initialValue: 100)}
            Center(
              child: Text('Fold : ${DataHelper.fold([10, 2, 5, 0.5], initialValue: 100)}\n'),
            ),

            ///skip
            Center(
              child: Text('Skip : ${DataHelper.skip([10, 2, 5, 0.5], 2)}\n'),
            ),

            ///take
            Center(
              child: Text('take : ${DataHelper.take([10, 2, 5, 0.5], 4)}\n'),
            ),

            ///join
            Center(
              child: Text('join : ${DataHelper.join([10, 2, 5, 0.5], '-')}\n'),
            ),

            ///last
            Center(
              child: Text('last : ${DataHelper.last([10, 2, 5, 0.5])}\n'),
            ),

            ///first
            Center(
              child: Text('first : ${DataHelper.first([10, 2, 5, 0.5])}\n'),
            ),

            ///remove
            Center(
              child: Text('remove : ${DataHelper.remove([
                    {'abx': 1},
                    {'abx': 2},
                    {'abx': 3}
                  ], [
                    {'abx': 2}
                  ])}\n'),
            ),
          ],
        ),
      ),
    );
  }
}


android/src/main/kotlin/com/test_pl/test_pl/TestPlPlugin.kt
package com.test_pl.test_pl

import androidx.annotation.NonNull

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** TestPlPlugin */
class TestPlPlugin: FlutterPlugin, MethodCallHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private lateinit var channel : MethodChannel

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "test_pl")
    channel.setMethodCallHandler(this)
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    if (call.method == "getPlatformVersion") {
      result.success("Android ${android.os.Build.VERSION.RELEASE}")
    } else {
      result.notImplemented()
    }
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }
}


lib/platform_version_helper.dart
import 'dart:async';
import 'dart:core';
import 'package:flutter/services.dart';

class DataHelper {
  static const MethodChannel _channel = MethodChannel('test_pl');
  /// getPlatformVersion
  static Future<dynamic?> get platformVersion async {
    final String? version = await _channel.invokeMethod('getPlatformVersion');
    return version;
  }
  /// isNullOrBlank
  static bool isNullOrBlank(dynamic val) {
    if (val is List) {
      if (val.isEmpty) return true;
        return false;
     } else {
      if (val == null || val == '' || val.toString().isEmpty) return true;
        return false;
    }
  }
  /// getJsonValue
  static Future<dynamic> getJsonValue(json, String path, defaultValue) {
    try {
      List pathSplitter = path.split(".");

      /// <String,dynamic> || String
      dynamic returnValue;

      json.forEach((key, value) {
        var currentPatten = pathSplitter[0];
        int index = 0;

        if (currentPatten.contains('[') && currentPatten.contains(']')) {
          int index1 = currentPatten.indexOf('[');
          int index2 = currentPatten.indexOf(']');

          index = int.parse(currentPatten.substring(index1 + 1, index2));
          currentPatten = currentPatten.substring(0, index1);
        }

        if (key == currentPatten) {
          if (pathSplitter.length == 1) {
            returnValue = value;
            return;
          }

          pathSplitter.remove(pathSplitter[0]);

          if (value == null) {
            returnValue = defaultValue;
            return;
          }
          try {
            try {
              value = value.toJson();
            } catch (error) {
              // handle error
            }

            try {
              if (value is List) {
                value = value[index];
              }
            } catch (error) {
              // handle error
            }

            returnValue = getJsonValue(value, pathSplitter.join("."), defaultValue);
          } catch (error) {
            returnValue = defaultValue;
          }
          return;
        }
      });

      return returnValue ?? defaultValue;
    } on Exception {
      // TODO
      return defaultValue;
    }

    //ex : helper.jsonGet(jobDetailCtrl.jobData, "salary_range", null);
    //ex : helper.jsonGet(jobDetailCtrl.jobData, "salary_range.from", '');
    //ex : helper.jsonGet(jobDetailCtrl.jobData, "salary_range.from.amount_gross", 0);
    //ex : helper.jsonGet(jobDetailCtrl.jobData, "salary_range[0].from.amount_gross", 'null');
    //ex : helper.jsonGet(jobDetailCtrl.jobData, "salary_range[0].from[1].amount_gross", null);
  }
  /// arrayFilter
  static List arrayFilter(List val) {
    if (val.isNotEmpty) {
      List newArray = [];
      for (int i = 0; i < val.length; i++) {
        if (val[i] != null) {
          newArray.add(val[i]);
        }
      }
      return newArray;
    } else {
      return [];
    }
  }
  /// chunk
  static chunk(array, [size = 1]) {
    var chunks = [];
    int chunkSize = 2;
    for (var i = 0; i < array.length; i += chunkSize) {
      chunks.add(array.sublist(i, i + chunkSize > array.length ? array.length : i + chunkSize));
    }
    return chunks;
  }
  /// isInteger
  static bool isInteger(value) => int.tryParse(value.toString()) != null ? true : false;
  /// isNumeric
  static bool isNumeric(s) {
    if (s == null) {
      return false;
    }
    return double.tryParse(s) != null;
  }
  /// toNumber
  static toNumber(value) {
    final type = value.runtimeType;
    if(type is String) {
      return double.tryParse(value);
    }
    return value;
  }
  /// compact
  static List compact(array) {
    // final myList = [10, '', '', null, 20, 0, 40, 50, false];

    // remove falsy values from the list
    array.removeWhere((item) => ["", null, false, 0].contains(item));
    return array;
  }
  /// concat
  static concat(array1, array2) {
    if(array2 is List) {
      array1.addAll(array2);
    } else {
      array1.add(array2);
    }
    return array1;
  }
  /// differenceFromFirst
  static differenceFromFirst(first, second) {
    List<dynamic> difference = first.toSet().difference(second.toSet()).toList();
    return difference;
  }
  /// difference
  static difference(l1,l2) {
    List<T> listDiff<T>(List<T> l1, List<T> l2) => (l1.toSet()..addAll(l2))
        .where((i) => !l1.contains(i) || !l2.contains(i))
        .toList();
    return listDiff(l1, l2);
  }
  /// removeAt
  static removeAt(array, index){
    array.removeAt(index);
    return array;
  }
  /// findIndex
  static findIndex(array, val){
    final index;
    if(val is String || val is int || val is Uri || val is DateTime){
      index = array.indexWhere((element) => element == val);
    } else {
      bool match = false;
      List keys = val.keys.toList();
      index = array.indexWhere((element) {
        for(var i=0;i<keys.length;i++) {
          if(i != 0 && !match){
            match = false;
            break;
          }
          element[keys[i]] == val[keys[i]] ? match = true : match = false;
        }
        return match;
      });
    }
    return index;
  }
  /// findLastIndex
  static findLastIndex (array, val){
    final index;
    if(val is String || val is int || val is Uri || val is DateTime){
      index = array.lastIndexWhere((element) => element == val);
    } else {
      bool match = false;
      List keys = val.keys.toList();
      index = array.lastIndexWhere((element) {
        for(var i=0;i<keys.length;i++) {
          if(i != 0 && !match){
            match = false;
            break;
          }
          element[keys[i]] == val[keys[i]] ? match = true : match = false;
        }
        return match;
      });
    }
    return index;
  }
  /// contains
  static contains (array, val){
    if(!isNullOrBlank(array)) return array.contains(val);
    return false;
  }
  ///containsfromList
  static containsfromList (array, val){
    if(!isNullOrBlank(array)) return array.values.contains(val);
    return null;
  }
  ///fold
  static fold (array, {initialValue = 0}){
    if(!isNullOrBlank(array)) return array.fold<dynamic>(
        initialValue, (previousValue, element) => previousValue + element);
    return array;
  }
  ///skip
  static skip (array, skipVal){
    if(!isNullOrBlank(array)) return array.skip(skipVal);
    return array;
  }
  ///take
  static take (array, takeVal){
    if(!isNullOrBlank(array)) return array.take(takeVal);
    return array;
  }
  ///join
  static join (array, takeVal){
    if(!isNullOrBlank(array)) return array.join(takeVal);
    return array;
  }
  ///last
  static last (array){
    if(!isNullOrBlank(array)) return array.last;
    return array;
  }
  ///first
  static first (array){
    if(!isNullOrBlank(array)) return array.first;
  }
  ///isEqual
  static isEqual (obj1, obj2){
    // return expect(obj1 == obj2, false);
  }
  ///remove
  static remove (List array, List val){
    if(val is List) {
      for(var obj in val){
        array.remove(obj);
      }
      print(array);
      return array;
    }else {
      array.removeAt(findIndex(array, val));
      return array;
    }
  }


}

// final DataHelper dataHelper = DataHelper();
